metadata/program/src/utils.rs

```diff
@@ -0,0 +1,65 @@
+use crate::error::MetadataError;
+use solana_program::{
+    account_info::AccountInfo,
+    entrypoint::ProgramResult,
+    program::invoke_signed,
+    program_error::ProgramError,
+    program_pack::{IsInitialized, Pack},
+    pubkey::Pubkey,
+    system_instruction::create_account,
+    sysvar::rent::Rent,
+};
+
+/// assert rent exempt
+pub fn assert_rent_exempt(rent: &Rent, account_info: &AccountInfo) -> ProgramResult {
+    if !rent.is_exempt(account_info.lamports(), account_info.data_len()) {
+        Err(MetadataError::NotRentExempt.into())
+    } else {
+        Ok(())
+    }
+}
+
+/// assert ununitialized account
+pub fn assert_uninitialized<T: Pack + IsInitialized>(
+    account_info: &AccountInfo,
+) -> Result<T, ProgramError> {
+    let account: T = T::unpack_unchecked(&account_info.data.borrow())?;
+    if account.is_initialized() {
+        Err(MetadataError::AlreadyInitialized.into())
+    } else {
+        Ok(account)
+    }
+}
+
+/// assert initialized account
+pub fn assert_initialized<T: Pack + IsInitialized>(
+    account_info: &AccountInfo,
+) -> Result<T, ProgramError> {
+    let account: T = T::unpack_unchecked(&account_info.data.borrow())?;
+    if !account.is_initialized() {
+        Err(MetadataError::Uninitialized.into())
+    } else {
+        Ok(account)
+    }
+}
+
+/// Create account from scratch, stolen from Wormhole, slightly altered for my purposes
+/// https://github.com/bartosz-lipinski/wormhole/blob/8478735ea7525043635524a62db2751e59d2bc38/solana/bridge/src/processor.rs#L1335
+#[inline(always)]
+pub fn create_account_raw<T: Pack>(
+    accounts: &[AccountInfo],
+    new_account: &Pubkey,
+    payer: &Pubkey,
+    owner: &Pubkey,
+    seeds: &[&[u8]],
+) -> Result<(), ProgramError> {
+    let size = T::LEN;
+    let ix = create_account(
```

Comment:
It's probably time to bust out https://github.com/solana-labs/solana-program-library/blob/7d4873c61721aca25464d42cc5ef651a7923ca79/associated-token-account/program/src/processor.rs#L51-L98 into a helper.

As this code is currently written, I can grief your NTF Metadata by running `solana transfer <metadata_address> 0.001` and prevent you from creating metadata

Reply:
This address can only be created by the program itself as it's a PDA, and only created within it and by it, since it must be signer, so I'm not sure who would be doing the grief-ing here. What are you suggesting - that we raise the minimum balance? I'm not entirely following.

Reply:
Nevermind, I understand what you mean. Will fix.

Reply:
Yeah it's kinda annoying the SystemInstruction::CreateAccount fails if the account already exists, and you can SystemInstruction::Transfer into any account to create it!